add nodouble option for all cwise ops

guoyejun opened the pull request:
this is the following patch of ac98d1184008e4 to support nodouble
option for all cwise ops. The macros REGISTER* defined within
__ANDROID_TYPES_SLIM__ should be changed to empty, and it impacts
all the cwise ops, so, all the changes for the cwise ops have to be
in a single patch.

there will be more patches for other ops to support nodouble option.

guoyejun committed:
add nodouble option for all cwise ops

this is the following patch of ac98d1184008e4 to support nodouble
option for all cwise ops. The macros REGISTER* defined within
__ANDROID_TYPES_SLIM__ should be changed to empty, and it impacts
all the cwise ops, so, all the changes for the cwise ops have to be
in a single patch.

there will be more patches for other ops to support nodouble option.

tensorflow-jenkins commented:
Can one of the admins verify this patch?

vrv commented on tensorflow/core/kernels/cwise_ops_common.h:
```diff
@@ -476,17 +476,17 @@ struct ApproximateEqual<CPUDevice, T> {
-  REGISTER(OP, D, N, F, T0)
-#define REGISTER9(OP, D, N, F, T0, T1, T2, T3, T4, T5, T6, T7, T8) \
-  REGISTER(OP, D, N, F, T0)
+// the float data type is registered with TF_CALL_SYCL_NUMBER_TYPES,
```
I don't really like that we're relying on a SYCL define for ANDROID_TYPES_SLIM, which are two different systems.  Can we make it so that SYCL and ANDROID_TYPES_SLIM don't interfere in any way?

guoyejun replied:
thanks for pointing it, i realized that i had a misunderstanding that REGISTER* is only for sycl backend, actually, it is for all the backends cpu, cuda and sycl.  My change relative to ANDROID_TYPES_SLIM is not right. I'll send new patches.

guoyejun commented:
thanks @vrv pointing my mistake, i'll rewrite the code to add new macros such as REGISTER_SYCL, and REGISTER6_SYCL. In the new code, the change could be separated to be one patch for one op, it could make each patch simple and easy to be reviewed.  So, my plan is to close this PR, and send out new PRs one by one (send out the second patch only after the first one is accepted).

vrv commented:
I'd be fine with all the changes in the same PR actually, since they are logically all the same, and can be tested all at once (less load on our test infrastructure). They are mostly mechanical changes so it's easier to review in a single batch :)

guoyejun commented:
got it, i'll continue this PR with new patches.

guoyejun committed:
add new macros REGISTER*_SYCL

When __ANDROID_TYPES_SLIM__ is defined, the sycl kernel register
is different with REGISTER*, so add new macros REGISTER*_SYCL.

rmlarsen commented:
@vrv Could you take another look, please?

vrv commented:
I think @benoitsteiner probably wants to look at this since it's SYCL related.

rmlarsen commented:
@tensorflow-jenkins test this please

guoyejun commented:
hello, i checked the failure log but do not understand how it relative to my change, could anyone help? thanks.

the last few lines of the log:
PASS: //tensorflow/python:session_benchmark
PASS: //tensorflow/python:matmul_benchmark_test
Build was aborted
Aborted by unknown
[Set GitHub commit status (universal)] PENDING on repos [] (sha:ef4e31e) with context:tensorflow-pull-requests-gpu
Unable to get pull request builder trigger!!
Setting status of 12276a750d6240f93878f84fa125f6be48566e7d to FAILURE with url https://ci.tensorflow.org/job/tensorflow-pull-requests-gpu/6269/ and message: 'FAILURE
 '
Using context: Linux GPU
Finished: ABORTED

rmlarsen commented:
@guoyejun The failure was unrelated to your change.

rmlarsen commented:
@tensorflow-jenkins test this please

benoitsteiner commented on tensorflow/core/kernels/cwise_op_abs.cc:
```diff
@@ -38,7 +38,11 @@ REGISTER_KERNEL_BUILDER(Name("Abs")
 
 #if TENSORFLOW_USE_SYCL
-REGISTER3(UnaryOp, SYCL, "Abs", functor::abs, float, double, int64);
+#define REGISTER_KERNEL(type) REGISTER(UnaryOp, SYCL, "Abs", functor::abs, type);
```
Can you create a new macro that call # define REGISTER_KERNEL(type) TF_CALL_SYCL_NUMBER_TYPES(REGISTER_KERNEL) in a central location and also make sure that the new macro undefines REGISTER_KERNEL when it's done. That will avoid a lot of boilerplate code.

guoyejun replied:
yes, good idea.

guoyejun replied:
i tried it, but did not found a way to implement it. My code looks like below, while it is illegal grammar.  Do you have any idea @benoitsteiner , thanks.

#define REGISTER_NUMBER_SYCL(OP, N, F)          \
  #define REGISTER_SYCL_KERNEL(type) REGISTER(OP, SYCL, N, F, type)   \
  TF_CALL_SYCL_NUMBER_TYPES(REGISTER_SYCL_KERNEL);                    \
  #undef REGISTER_SYCL_KERNEL

benoitsteiner commented on tensorflow/core/kernels/cwise_ops_sycl_common.h:
```diff
@@ -154,6 +154,28 @@ struct BinaryFunctor<SYCLDevice, Functor, NDIMS, has_errors> {
 }  // end namespace functor
+
+
+#if defined(__ANDROID_TYPES_SLIM__)
```
Why is this needed for coefficient wise operations and not for all the other sycl kernels ?

guoyejun replied:
In the current code, it is needed for cwise ops, see the code at the end of cwise_ops_common.h,  this change is to keep this logic.

As for other sycl kernels, i'll review it when I face them. As for this patch, it just focuses on cwise ops.

benoitsteiner reviewed (changes requested):
I'm not sure this is a scalable approach.

martinwicke commented:
@benoitsteiner Can you take a look at the replies?

drpngx commented:
@benoitsteiner what do you think?

sb2nov commented:
@benoitsteiner ping?

drpngx commented:
@benoitsteiner any luck with this?

tensorflowbutler commented:
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

tensorflowbutler commented:
Nagging Assignee @benoitsteiner: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

tensorflowbutler commented:
Nagging Assignee @benoitsteiner: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

guoyejun commented:
just close it since no feedback for a long time, and code base change results into conflicts
